### ISSUE TITLE
Invalid argument at /lib/LANraragi/Model/Upload.pm line 158.

### ISSUE THREAD

**Issue by Vhu3** [Difegue/LANraragi#626]
**LRR Version and OS**
Windows, Lanraragi v8.5

**Bug Details**
When inserting multiple Download URLs to request from the downloader, some of the downloads fail. The download does not fail if i spam the request to the server. I fail to consistently reproduce the issue. 

This is the line that says there is an invalid argument "my $content_disp = $tx->result->headers->content_disposition;".

Not sure if this would be the issue, but i am using my own downloader plugin, here is the sub if it helps. 
sub provide_url {
    shift;
    my $lrr_info = shift;

    # Get the URL to download
    my $url = $lrr_info->{url};
    my $ua = $lrr_info->{user_agent};
    my $res = $ua->max_redirects(5)->get($url)->result->dom;
    my $dl = $res->at('a[href$="/download"]')->attr('href');

    return ( download_url => $dl);
}

**Matching Logs**
[Minion] [info] Done!
Mojo::Reactor::Poll: I/O watcher failed: Can't call method "emit" on an undefined value at /usr/local/share/perl5/site_perl/Mojo/Redis.pm line 60.
Mojo::Reactor::Poll: I/O watcher failed: Can't call method "emit" on an undefined value at /usr/local/share/perl5/site_perl/Mojo/Redis.pm line 60.
[Minion] [info] Warming up search cache...
[2022-05-04 14:14:54.20800] [53] [info] Creating process id file "/home/koyomi/lanraragi/public/temp/server.pid"
[2022-05-04 14:14:54.22888] [53] [info] Creating process id file "/home/koyomi/lanraragi/public/temp/server.pid"
[2022-05-04 14:14:55.23040] [53] [info] Creating process id file "/home/koyomi/lanraragi/public/temp/server.pid"
[2022-05-04 14:14:56.23182] [53] [info] Creating process id file "/home/koyomi/lanraragi/public/temp/server.pid"
[Minion] [info] Done!
[Minion] [info] Downloading url https://koushoku.org/archive/294...
[Minion] [info] Found downloader koushoku
[Plugin System] [info] No login plugin specified, returning empty UserAgent.
[Minion] [info] URL transformed by plugin to https://archive/294/magical-lab/download
[File Upload/Download] [info] Downloading URL https://archive/294/magical-lab/download...This will take some time.
[Minion] [info] URL downloaded to /tmp/oA9FnuHjhc/[Kamiya Zuzu].zip
[Auto-Plugin] [info] Executing enabled metadata plugins on archive with id 084183efb7f856afa3e52e6390a8433b0b3446d2.
Mojo::Reactor::Poll: I/O watcher failed: Can't call method "emit" on an undefined value at /usr/local/share/perl5/site_perl/Mojo/Redis.pm line 60.
Mojo::Reactor::Poll: I/O watcher failed: Can't call method "emit" on an undefined value at /usr/local/share/perl5/site_perl/Mojo/Redis.pm line 60.
[Minion] [info] Warming up search cache...
[2022-05-04 14:15:36.23268] [53] [info] Creating process id file "/home/koyomi/lanraragi/public/temp/server.pid"
[2022-05-04 14:15:37.23414] [53] [info] Creating process id file "/home/koyomi/lanraragi/public/temp/server.pid"
[Minion] [info] Done!

**Comment by Difegue** [comment]
There's been some issues with Windows + 0.8.5 due to underlying changes, notably inconsistent results with network requests -- If you're using WSL1, can you try switching to WSL2 and see if that improves the issue?  

I'm not seeing anything wrong in that custom downloader plugin at a glance. 👍

**Comment by Vhu3** [comment]
Sadly I cannot upgrade to WSL2 Since that would disable the usage of VMs that i currently use.

**Comment by Difegue** [comment]
I understand; Does the issue happen on 0.8.4 as well, then?

**Comment by Vhu3** [comment]
Interesting. Whatever Changed between 0.8.4 and 0.8.5 broke the downloader. I was able to download a batch of 100 without having any issues on 0.8.4 but on 0.8.5 i can barley download 5 downloads before i get the errors described above. Thank for downgrade tip. Just started using LANraragi. Note both were still ran on WSL1 as i do not have a way to test WSL2.

**Comment by Difegue** [comment]
Yeah I'm guessing this is the same issue as #619 / #621 / #622 .
0.8.5 switched from Alpine Linux 3.12 to 3.14 as a base, which can break DNS lookups for websites which do not provide a AAAA ipv6 record: See https://gitlab.alpinelinux.org/alpine/aports/-/issues/11879 ,  https://github.com/home-assistant/core/issues/57173 or https://github.com/docker/for-mac/issues/5088 .

I might try and introduce a dedicated DNS lib to mitigate this issue, (cf. https://christoph.luppri.ch/fixing-dns-resolution-for-ruby-on-alpine-linux ) but in the meantime users can either downgrade or use WSL2.

**Comment by anonVin79513** [comment]
Nope. It did not make things any better. If anything I think more downloads in a batch are failing now.

**Comment by Difegue** [comment]
I've had reports that the `Net::DNS::Native` switch does improve things for the few Linux hosts also encountering this problem, so that's probably going to stay in for now.  

As for Windows+WSL1, I'm still investigating potential solutions: One of them seems to be **disabling IPv6** on the Windows host.  

Past that, it's unlikely WSL1 will receive updates by Microsoft to fix the issue, so I'll have to either:  
- Build legacy images on alpine 3.12 and use those in Windows builds -- This needs some extra CI/workflow changes and feels like a bandaid tbh
- Use custom DNS resolving in the entire app -- There's no easy custom DNS resolver for mojolicious so that's probably the most expensive solution
- Go all in on WSL2 by notably fixing its biggest flaw through #580 -- This is the easiest and fixes an existing problem as well, but doesn't fix the issue for users that are stuck on WSL1.

**Comment by Difegue** [comment]
I went for the bandaid in the end :^)  

Other options are too much of a pain and I can't decently drop WSL1 yet since v2 doesn't cover all bases even with the hostpatcher.   

Windows nightlies now use alpine 3.12 again -- There's no change to the regular Docker images but you can build a 3.12 image if you want by using `Dockerfile-legacy` in /tools.
